### PR TITLE
8297286: runtime/vthread tests crashing after JDK-8296324

### PR DESCRIPTION
**Pull request by sspitsyn**
This problem has two sides.
One is that the `VirtualThread::run() `cashes the field `notifyJvmtiEvents` value.
It caused the native method `notifyJvmtiUnmountBegin()` not called after the field `notifyJvmtiEvents`
value has been set to `true` when an agent library is loaded into running VM.
The fix is to get rid of this cashing.
Another is that enabling `notifyJvmtiEvents` notifications needs a synchronization.
Otherwise, a VTMS transition start can be missed which will cause some asserts to fire.
The fix is to use a JvmtiVTMSTransitionDisabler helper for sync.

Testing:
The originally failed tests are passed now:
```
runtime/vthread/RedefineClass.java
runtime/vthread/TestObjectAllocationSampleEvent.java 
```
In progress:
Run the tiers 1-6 to make sure there are no regression.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [ ] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Integration blocker
&nbsp;⚠️ Whitespace errors (failed with the updated jcheck configuration)

### Issue
 * [JDK-8297286](https://bugs.openjdk.org/browse/JDK-8297286): runtime/vthread tests crashing after JDK-8296324

### Reviewers
 * [Leonid Mesnik](https://openjdk.org/census#lmesnik) (@lmesnik - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/11304/head:pull/11304` \
`$ git checkout pull/11304`

Update a local copy of the PR: \
`$ git checkout pull/11304` \
`$ git pull https://git.openjdk.org/jdk.git pull/11304/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11304`

View PR using the GUI difftool: \
`$ git pr show -t 11304`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11304.diff">https://git.openjdk.org/jdk/pull/11304.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/11304#issuecomment-1324402739)

### CODE REVIEW

File: src/java.base/share/classes/java/lang/VirtualThread.java [hunk 1]
```diff
@@ -270,11 +270,10 @@ private void lazySubmitRunContinuation() {
     @ChangesCurrentThread
     private void run(Runnable task) {
         assert state == RUNNING;
-        boolean notifyJvmti = notifyJvmtiEvents;
```
Comment (lmesnik) [inline src/java.base/share/classes/java/lang/VirtualThread.java:273]: Don't we have same issue in yieldContinuation() method? (line 396)

**Review by lmesnik — APPROVED** [review]

**Review by lmesnik — CHANGES_REQUESTED** [review]
Needed to check yieldContinuation() method for same issue.

File: src/java.base/share/classes/java/lang/VirtualThread.java (hunk 1 above)
Comment (sspitsyn) [reply src/java.base/share/classes/java/lang/VirtualThread.java:273]: Fixed the `yieldContinuation()` method.
There is also `switchToCarrierThread()` method that returns the `notifyJvmtiEvents` value.
It seems to be an optimization. I'm not sure yet, if we need to fix these places as well.

Comment (AlanBateman) [reply src/java.base/share/classes/java/lang/VirtualThread.java:273]: > There is also `switchToCarrierThread()` method that returns the `notifyJvmtiEvents` value.
> It seems to be an optimization. I'm not sure yet, if we need to fix these places as well.

It was to ensure that hide(true) and hide(false) are balanced. If it were to re-poll notifyJvmtiEvents and a JVMTI agent enables the capability while a thread in doing a temporary transition then you may get a hide(false) without the corresponding hide(true).

Comment (sspitsyn) [reply src/java.base/share/classes/java/lang/VirtualThread.java:273]: Okay, I see it now.

**Review by lmesnik — APPROVED** [review]

**Comment by AlanBateman** [comment]
Would it be possible to summarize behavior for when an agent enables the capability as a virtual thread executes for the first time or it continues after yield? More specifically JVMTI will be notified of a mount end without a correspond mount begin. It might be that we can narrow this down to if finish_VTMS_transition is okay without a preceding start_VTMS_transition.

**Comment by sspitsyn** [comment]
I've forgotten the `JvmtiVTMSTransitionDisabler` is not going to work before the `notifyJvmtiEvents` is set to `true`.
I agree, we may want to allow start_VTMS_transition/finish_VTMS_transition not properly paired as you suggest.
But then it is not good that we loose the ability to strictly check/assert VTMS transition notifications for other cases.
Need to think a bit more on this.

**Comment by sspitsyn** [comment]
This is artificial comment to keep the PR alive.

**Comment by sspitsyn** [comment]
Need to keep this PR alive for a while.

**Pull request closed without merging**